### PR TITLE
Refactor NYPL script to use ProviderDataIngester class

### LINKED ISSUE #590: Refactor nypl to use ProviderDataIngester
## Suggested Improvement

Refactor the nypl provider script to use the new ProviderDataIngester base class

## Benefit

More details in #229

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### LINKED ISSUE #710: NYPL: `physical_description` might be `None`
## Description

We encountered this error with the most recent run of the NYPL DAG:

```
[2022-09-01, 00:00:06 UTC] {factory_utils.py:135} INFO - Beginning ingestion
[2022-09-01, 00:00:09 UTC] {nypl.py:159} WARNING - No primary creator found
[2022-09-01, 00:00:09 UTC] {taskinstance.py:1909} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 171, in execute
    return_value = self.execute_callable()
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 189, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/usr/local/airflow/openverse_catalog/dags/providers/factory_utils.py", line 140, in pull_media_wrapper
    data = run_func(*args)
  File "/usr/local/airflow/openverse_catalog/dags/providers/provider_api_scripts/nypl.py", line 49, in main
    _handle_results(results)
  File "/usr/local/airflow/openverse_catalog/dags/providers/provider_api_scripts/nypl.py", line 105, in _handle_results
    metadata = _get_metadata(mods)
  File "/usr/local/airflow/openverse_catalog/dags/providers/provider_api_scripts/nypl.py", line 205, in _get_metadata
    if description := physical_description.get("note", {}).get("$"):
AttributeError: 'NoneType' object has no attribute 'get'
```

## Additional context

Airflow log for those with access: https://airflow.openverse.engineering/log?execution_date=2022-08-01T00%3A00%3A00%2B00%3A00&task_id=pull_image_data&dag_id=nypl_workflow&map_index=-1

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #590 by @stacimc, fixes #710 by @AetherUnbound 

## Description

This PR replaces #546. Changes in the NYPL script:
- use of the DataIngester class. NYPL requests a single image API endpoint to get each image's data. To enable delayed requests for _detail_ endpoints, I added `get_detail_json` method to the base class, but I think its API can be greatly improved.
- filetype detection.
- category is set as `photograph`  for items with `photograph` as their `genre`.
- creator is added when there is a single creator, not only when there's an array of creators.
- `date_created` added to `meta_data`.
- `topics` extracted and added to `meta_data`. I was wondering if we should combine `physicalDescription` and `topics` into a single string and save it as `description` to improve the search relevancy for NYPL.

## Testing Instructions

Run NYPL DAG - it should run without errors. You can stop the `pull_data` step early to review the data that is saved in the database.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:25]: Should this maybe be the following for reuseability?
```suggestion
    endpoint = "http://api.repo.nypl.org/api/v1/items"
    search_endpoint = f"{endpoint}/search"
    metadata_endpoint = f"{endpoint}/item_details/"
```

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:90]: I think we can just use `self` here:
```suggestion
            image_url, filetype = self._get_image_data(image_link)
```

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:200]: I'm not sure this will work - `start_date` and `end_date` are generators so this will just interpolate the generator representation into `date_created` 😮 Do we have a test for this particular code path?

File: openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py:180]: Since this is almost exactly the same as `get_response_json`, do you think we could just use that instead and pass in `{}` if necessary in the operational code?

**Review by AetherUnbound — CHANGES_REQUESTED** [review]
Yay for refactors! I have one note regarding generators that I think will be problematic, and another regarding function re-use.

File: openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py
Comment (AetherUnbound) [reply openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py:180]: Ah, I'm noticing now that this allows the specification of an endpoint! I also have a use for this in #583 - I'll see if I can still fold this into the `get_response_json` function in my PR!

Comment (AetherUnbound) [reply openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py:180]: Here's how I plan on modifying it on my branch:

```diff
diff --git a/openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py b/openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py
index 8d9d2efc..adf19295 100644
--- a/openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py
+++ b/openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py
@@ -315,13 +315,16 @@ class ProviderDataIngester(ABC):
 
         return batch, should_continue
 
-    def get_response_json(self, query_params: dict):
+    def get_response_json(self, query_params: dict, endpoint: str | None = None):
         """
         Make the actual API requests needed to ingest a batch. This can be overridden
         in order to support APIs that require multiple requests, for example.
         """
         return self.delayed_requester.get_response_json(
-            self.endpoint, self.retries, query_params, headers=self.headers
+            endpoint or self.endpoint,
+            self.retries,
+            query_params,
+            headers=self.headers,
         )
 
     def get_should_continue(self, response_json):
```

Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py:180]: This is great, thank you @AetherUnbound !

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (stacimc) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:221]: Do these make sense as tags?

**Review by stacimc — CHANGES_REQUESTED** [review]
Once the `endpoint` is updated, this is running pretty well for me locally, just the couple of errors!

I'm really excited about all of the additions, but I will say that I'm concerned with having so many extra changes bundled up in these refactors. Generally refactors should have as few changes to functionality as possible.

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [reply openverse_catalog/dags/providers/provider_api_scripts/nypl.py:25]: Oops, I missed that! Sorry Olga, good to know going forward!

**Comment by krysal** [comment]
@obulat Is something blocking this PR? Looks like is a WIP given that some tests are failing so I'll convert it to draft in the meantime.

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:124]: Maybe we could add this or something similar?
```suggestion
        "Cropped .jpeg (1600 pixels on the long side)"
        This is required because the filetype is not present/extractable from the
        url via the MediaStore class.
```

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:135]:
```suggestion
        Receives a list of dictionaries of the following shape:
```

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:133]: Suggesting types to more clearly see what this function returns
```suggestion
    def _get_image_data(images) -> tuple[None, None] | tuple[str, str]:
```

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:158]: If none of the possible dimensions are available, this will raise a `StopIteration` exception:
```python
[ins] In [2]: next(x for x in [])
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ <ipython-input-2-dbd4591bda43>:1 in <module>                                                     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
StopIteration
```

Does that have a possibility of occurring? Should we construct this as a list then see if it has any length rather than using `next`?

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (AetherUnbound) [inline openverse_catalog/dags/providers/provider_api_scripts/nypl.py:227]: How annoying!

File: tests/dags/providers/provider_api_scripts/test_nypl.py
Comment (AetherUnbound) [inline tests/dags/providers/provider_api_scripts/test_nypl.py:23]: This can also be removed 😄 
```suggestion
```

File: tests/dags/providers/provider_api_scripts/test_nypl.py
Comment (AetherUnbound) [inline tests/dags/providers/provider_api_scripts/test_nypl.py:35]: Rather than initializing a new media store, maybe we could use the one that comes with the ingester?
```suggestion
image_store = nypl.media_stores["image"]
```

File: tests/dags/providers/provider_api_scripts/test_nypl.py
Comment (AetherUnbound) [inline tests/dags/providers/provider_api_scripts/test_nypl.py:147]: Pytest should be able to show a reasonable diff for this output, I think we can just do (unless there are more fields present in `image[0]` we don't want to check against):
```suggestion
    assert images[0] == expected_image
```

**Review by AetherUnbound — COMMENTED** [review]
I have a few more thoughts, this is really close!

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/nypl.py]: Thank you, I didn't know this! I've refactored the code here.

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/nypl.py:254]: It is everywhere in the API replies :(
I've added a helper to simplify some usecases.

**Comment by obulat** [comment]
I've fixed the code to get the "physical description", but I'm not sure that we really need it. Here's one example:
"4 polyester film encapsulations, some containing 2 sheets back-to-back. Accompanying text formatted as 1 large sheet (46 x 59 cm), in one of the encapsulations." - doesn't seem to add a lot to the search quality.

**Comment by AetherUnbound** [comment]
> I've fixed the code to get the "physical description", but I'm not sure that we really need it. Here's one example: "4 polyester film encapsulations, some containing 2 sheets back-to-back. Accompanying text formatted as 1 large sheet (46 x 59 cm), in one of the encapsulations." - doesn't seem to add a lot to the search quality.

I feel like from some museum responses I've seen things like "wood cutout x by y size" which could help for searches in some cases 😮 I think it's worth including!

**Review by AetherUnbound — APPROVED** [review]
Great work Olga! I have one request regarding the API key retrieval that I think would be important to have before merging. Otherwise I think this is ready! Tested locally and it ran fine 😄 🚀

**Review by stacimc — APPROVED** [review]
This is working locally for me :) Really nice, I like how you've handled the types.

File: openverse_catalog/dags/providers/provider_api_scripts/nypl.py
Comment (obulat) [reply openverse_catalog/dags/providers/provider_api_scripts/nypl.py:27]: Thank you! I've tried to find the way to split the list but didn't know how to search for it, apparently.

**Pull request merged**
